core/kernel/transfer_list.c

```diff
@@ -0,0 +1,524 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, Linaro Limited
+ */
+
+#include <kernel/cache_helpers.h>
+#include <kernel/panic.h>
+#include <kernel/transfer_list.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <string.h>
+#include <util.h>
+
+static void unmap_list(struct transfer_list_header *tl, size_t sz)
+{
+	if (core_mmu_remove_mapping(MEM_AREA_TRANSFER_LIST, tl, sz))
+		panic("Failed to remove transfer list mapping");
+}
+
+struct transfer_list_header *transfer_list_map(paddr_t pa)
+{
+	struct transfer_list_header *tl = NULL;
+	size_t sz = SMALL_PAGE_SIZE;
+	size_t old_sz = 0;
+
+	while (true) {
+		tl = core_mmu_add_mapping(MEM_AREA_TRANSFER_LIST, pa, sz);
+		if (!tl) {
+			EMSG("Failed to map TL with PA %#"PRIxPA", size %#zx",
+			     pa, sz);
+			return NULL;
+		}
+		old_sz = sz;
+
+		if (transfer_list_check_header(tl) == TL_OPS_NON) {
+			unmap_list(tl, sz);
+			return NULL;
+		}
+
+		if (tl->max_size <= sz)
+			return tl;
+
+		sz = ROUNDUP(tl->max_size, SMALL_PAGE_SIZE);
+		unmap_list(tl, old_sz);
+	}
+}
+
+void transfer_list_unmap_sync(struct transfer_list_header *tl)
+{
+	size_t sz = tl->max_size;
+
+	transfer_list_update_checksum(tl);
+	dcache_cleaninv_range(tl, sz);
+	unmap_list(tl, sz);
+}
+
+void transfer_list_unmap_nosync(struct transfer_list_header *tl)
+{
+	unmap_list(tl, tl->max_size);
+}
+
+void transfer_list_dump(struct transfer_list_header *tl)
+{
+	struct transfer_list_entry *te = NULL;
+	int i = 0;
+
+	if (!tl)
+		return;
+
+	DMSG("Dump transfer list:");
+	DMSG("signature  %#"PRIx32"", tl->signature);
+	DMSG("checksum   %#"PRIx8"", tl->checksum);
+	DMSG("version    %#"PRIx8"", tl->version);
+	DMSG("hdr_size   %#"PRIx8"", tl->hdr_size);
+	DMSG("alignment  %#"PRIx8"", tl->alignment);
+	DMSG("size       %#"PRIx32"", tl->size);
+	DMSG("max_size   %#"PRIx32"", tl->max_size);
+	while (true) {
+		te = transfer_list_next(tl, te);
+		if (!te)
+			break;
+
+		DMSG("Entry %d:", i++);
+		DMSG("tag_id     %#"PRIx16"", te->tag_id);
+		DMSG("hdr_size   %#"PRIx8"", te->hdr_size);
+		DMSG("data_size  %#"PRIx32"", te->data_size);
+		DMSG("data_addr  %#"PRIxVA"",
+		     (unsigned long)transfer_list_entry_data(te));
+	}
+}
+
+/*******************************************************************************
+ * Creating a transfer list in a reserved memory region specified
+ * Compliant to 2.4.5 of Firmware handoff specification (v0.9)
+ * Return pointer to the created transfer list or NULL on error
+ ******************************************************************************/
+struct transfer_list_header *transfer_list_init(void *pa, size_t max_size)
```

Comment:
It seems that `paddr_t pa` would be more suitable.

Reply:
Fixed in the fixup patches.